src/dataframe/dataframe.jl

```diff
@@ -759,9 +790,16 @@ Copy data frame `df`.
 If `copycols=true` (the default), return a new  `DataFrame` holding
 copies of column vectors in `df`.
 If `copycols=false`, return a new `DataFrame` sharing column vectors with `df`.
+
+`copy` preserves table level and column level metadata.
 """
 function Base.copy(df::DataFrame; copycols::Bool=true)
-    return DataFrame(copy(_columns(df)), copy(index(df)), copycols=copycols)
+    cdf = DataFrame(copy(_columns(df)), copy(index(df)), copycols=copycols)
+    _copy_metadata!(cdf, df)
+    for col in _names(cdf)
+        _copy_colmetadata!(cdf, col, df, col)
+    end
```

Comment:
the constructor in this case calls `copy` so we need to define it here.

Reply:
I must be missing something: don't we copy metadata her anyway? :-)

Reply:
We do not have any constructor that accepts passing metadata dictionaries.
We could add it, but it was not needed.

Also I cannot just `copy` the internal `colmetadata` dictionary of `df` if it is present, because it would be a shallow copy so then the `Dict`s stored in it that store column metadata would be shared by `cdf` and `df`.

What I can do is extract-out the `colmetadata` dictionary from `df` and iterate its key-value pairs. I will make this change. It will be faster if data frame has many columns but only some of them have metadata.

Reply:
This is what I have now done.